### PR TITLE
Fix for #2543 - take into account WithTenant()

### LINKED ISSUE #2543: [Bug] [M] WithTenant ignored, resulting in CA failures for multi-tenant apps
MSAL 4.29 with WAM on UWP

## Repro

```csharp
var options = new PublicClientApplicationOptions();
             options.ClientId = token.ClientId;                            // “110797d6-4a5e-4e58-a06d-f1bf3f3a8069”

            var app = PublicClientApplicationBuilder
                .CreateWithApplicationOptions(options)
                .WithTenantId(token.TenantId)                             //The tenantId passed here (“3cfc49f9-956e-4e4e-a1b6-e03368c2e448”)
                .WithAuthority(token.AuthorityEndpoint)          //Set to the “common” endpoint “https://login.microsoftonline.com/common”
                .WithExperimentalFeatures(true)
               .WithBroker(true)
              
               .Build();
```

Expected: AcquireToken* should work for user 
Actual: AcquireToken* fails because the authority passed to WAM ends in "common", instead of the tenanted one, causing issues with Conditional Access

**Workaround**

Use a tenanted authority instead, i.e. `https://login.microsoftonline.com/3cfc49f9-956e-4e4e-a1b6-e03368c2e448`

### PR DESCRIPTION
**Pull request by bgavrilMS**
Fixes #2543  .

**Changes proposed in this request**

**Testing**

unit

**Performance impact**

Extra logic added only in corner-case scenario.

### CODE REVIEW

File: src/client/Microsoft.Identity.Client/AppConfig/AbstractApplicationBuilder.cs [hunk 1]
```diff
@@ -499,25 +498,38 @@ internal virtual void Validate()
...
         }
 
         #region Authority
-        private void CreateAuthorityInfoFromEnums()
+        private void ResolveAuthority()
         {
             if (Config.AuthorityInfo != null)
             {
-                return;
+                if (!string.IsNullOrEmpty(Config.TenantId) &&
```
Comment (trwalke) [inline src/client/Microsoft.Identity.Client/AppConfig/AbstractApplicationBuilder.cs:511]: Does this also cover application configurations like "PublicClientApplicationOptions"?

**Comment by trwalke** [comment]
Do we need a manual test to ensure WAM gets the proper authority?

**Review by trwalke — APPROVED** [review]

File: src/client/Microsoft.Identity.Client/MsalError.cs [hunk 2]
```diff
@@ -793,6 +793,11 @@ public static class MsalError
         /// </summary>
         public const string AuthorityTypeMismatch = "authority_type_mismatch";
 
+        /// <summary>
+        /// <para>What happens?</para>You specified a tenant twice - once in WithAuthority() and once using WithTenant()
+        /// <para>Mitigation</para>Specify the tenant once.
+        /// </summary>
+        public const string AuthorityTenantSpecifiedTwice = "authority_type_mismatch";
```
Comment (SameerK-MSFT) [inline src/client/Microsoft.Identity.Client/MsalError.cs:800]: The value of the constant AuthorityTenantSpecifiedTwice is same as that of AuthorityTypeMismatch.

**Review by SameerK-MSFT — CHANGES_REQUESTED** [review]

File: src/client/Microsoft.Identity.Client/MsalError.cs (hunk 2 above)
Comment (bgavrilMS) [reply src/client/Microsoft.Identity.Client/MsalError.cs:800]: Good catch

File: src/client/Microsoft.Identity.Client/AppConfig/AbstractApplicationBuilder.cs (hunk 1 above)
Comment (bgavrilMS) [reply src/client/Microsoft.Identity.Client/AppConfig/AbstractApplicationBuilder.cs:511]: Seems so. I added a test to it.

File: src/client/Microsoft.Identity.Client/AppConfig/AbstractApplicationBuilder.cs [hunk 3]
```diff
@@ -499,25 +498,38 @@ internal virtual void Validate()
...
+                    Config.AuthorityInfo.AuthorityType == AuthorityType.Aad)
+                {
+                    AadAuthority aadAuthority = Authority.CreateAuthority(Config.AuthorityInfo) as AadAuthority;
+                    if (!aadAuthority.IsCommonOrganizationsOrConsumersTenant())
+                    {
+                        throw new MsalClientException(MsalError.AuthorityTenantSpecifiedTwice, "You specified a tenant twice - once in WithAuthority() and once using WithTenant().");
```
Comment (neha-bhargava) [inline src/client/Microsoft.Identity.Client/AppConfig/AbstractApplicationBuilder.cs:517]: Do we also throw an exception when the tenant specified twice is same?

**Review by neha-bhargava — APPROVED** [review]

**Review by pmaytak — APPROVED** [review]

File: tests/Microsoft.Identity.Test.Unit/AppConfigTests/PublicClientApplicationBuilderTests.cs [hunk 4]
```diff
@@ -594,6 +594,50 @@ public void WithClientCapabilitiesViaOptions()
...
+            var options = new PublicClientApplicationOptions();
+            options.ClientId = TestConstants.ClientId;            
+
+            var app1 = PublicClientApplicationBuilder
+                .CreateWithApplicationOptions(options)
+                .WithTenantId(TestConstants.TenantId)                            
+                .WithAuthority("https://login.microsoftonline.com/common")       
+               .Build();
+
+            Assert.AreEqual($"https://login.microsoftonline.com/{TestConstants.TenantId}/", app1.Authority);
```
Comment (pmaytak) [inline tests/Microsoft.Identity.Test.Unit/AppConfigTests/PublicClientApplicationBuilderTests.cs:610]: nit: `$"https://login.microsoftonline.com/{TestConstants.TenantId}/"` can be changed to `TestConstants.AuthorityTenant`

**Pull request merged**